compiler-core/src/type_/expression.rs

```diff
@@ -10,28 +10,131 @@ use crate::{
+            javascript: true,
+        }
+    }
+
+    pub fn javascript() -> SupportedTargets {
+        SupportedTargets {
+            erlang: false,
+            javascript: true,
+        }
+    }
+
+    pub fn erlang() -> SupportedTargets {
+        SupportedTargets {
+            erlang: true,
+            javascript: false,
+        }
+    }
+
+    pub fn from_target(target: Target) -> SupportedTargets {
+        match target {
+            Target::Erlang => SupportedTargets::erlang(),
+            Target::JavaScript => SupportedTargets::javascript(),
+        }
+    }
+
+    pub fn intersect(&self, targets: SupportedTargets) -> SupportedTargets {
+        SupportedTargets {
+            erlang: self.erlang && targets.erlang,
+            javascript: self.javascript && targets.javascript,
+        }
+    }
+
+    pub fn merge(&self, targets: SupportedTargets) -> SupportedTargets {
+        SupportedTargets {
+            erlang: self.erlang || targets.erlang,
+            javascript: self.javascript || targets.javascript,
+        }
+    }
+
+    pub fn add(&self, target: Target) -> SupportedTargets {
+        match target {
+            Target::Erlang => SupportedTargets {
+                erlang: true,
+                javascript: self.javascript,
+            },
+            Target::JavaScript => SupportedTargets {
+                erlang: self.erlang,
+                javascript: true,
+            },
+        }
+    }
+
+    pub fn supports(&self, target: Target) -> bool {
+        match target {
+            Target::Erlang => self.erlang,
+            Target::JavaScript => self.javascript,
+        }
+    }
+
+    pub fn supports_all_targets(&self) -> bool {
+        self.javascript && self.erlang
+    }
+
+    pub fn to_vec(self) -> Vec<Target> {
+        let SupportedTargets { erlang, javascript } = self;
+        match (erlang, javascript) {
+            (true, true) => vec![Target::Erlang, Target::JavaScript],
+            (true, _) => vec![Target::Erlang],
+            (_, true) => vec![Target::JavaScript],
+            (_, _) => vec![],
+        }
+    }
+}
+
 #[derive(Debug)]
 pub(crate) struct ExprTyper<'a, 'b> {
     pub(crate) environment: &'a mut Environment<'b>,
 
+    pub(crate) supported_targets: SupportedTargets,
+
     // Type hydrator for creating types from annotations
     pub(crate) hydrator: Hydrator,
+
+    external_supported_targets: SupportedTargets,
 }
 
 impl<'a, 'b> ExprTyper<'a, 'b> {
-    pub fn new(environment: &'a mut Environment<'b>) -> Self {
+    pub fn new(
+        environment: &'a mut Environment<'b>,
+        external_supported_targets: SupportedTargets,
+    ) -> Self {
         let mut hydrator = Hydrator::new();
+
         hydrator.permit_holes(true);
         Self {
             hydrator,
             environment,
+            // This will be narrowed down as the expression type is inferred
+            supported_targets: SupportedTargets::all(),
+            external_supported_targets,
```

Comment:
Why are these tracks independently? Seems they always get merged together before use so it can be simplified to a single value. This would also permit the `merge` method to be removed.

Reply:
I think we have to keep track of those separately in the `ExprTyper`: when narrowing I have to perform this check 
```rust
self.supported_targets.narrow(new_targets);
if self.environment.target_support == TargetSupport::Enforced
   && !self.supported_targets
      .merge(self.external_supported_targets)
      .supports(self.environment.target) {
  // error this function has no supported targets and we must fail!
}
```
Supported targets get narrowed down starting from `SupportedTargets::all()`, the external targets, however, never change: imagine you had a function with an external js implementation that uses an Erlang-only function in its body:
```gleam
@external(javascript, "foo", "bar")
pub fn foo() {
  erlang_only()
}
```
If you had a single `supported_targets` field it would be narrowed down to just `SupportedTargets::erlang()` and you'd loose track of the fact that this function can actually support javascript as well. Now If you're compiling to JavaScript the check would fail:
```rust
self.supported_targets.narrow(new_targets);
if self.environment.target_support == TargetSupport::Enforced
   && !self.supported_targets.supports(self.environment.target) {
  // error this function has no supported targets and we must fail!
```
Because the body's supported target was narrowed down to just `Erlang` and you're not considering that you can still support JavaScript since you have an external implementation for that target.

Though the check could be written in a different way to make things a bit clearer and not rely on the merge function:

```rust
self.supported_targets.narrow(new_targets);
if self.environment.target_support == TargetSupport::Enforced
   && !(
     self.supported_targets.supports(self.environment.target)
     || self.external_targets.supports(self.environment.target)
   ) {
  // error this function has no supported targets and we must fail!
```

We would still need `merge` to merge the external targets and the body supported targets once the typer did its thing and record all the targets supported by the function, both those of its body and those with and external implementation.

Reply:
Ah I get it now, and you can't do that I originally said because that wouldn't give you the error at the point the non-supporting value is referenced. Thank you